### PR TITLE
[4.0] [RFC] Allow user to add a custom positions in the template

### PR DESCRIPTION
**Pull request by Fedik**
This is as an alternative to #19062

Quick and dirty example to what I wrote in that comment https://github.com/joomla/joomla-cms/pull/19062#issuecomment-355746742

### Summary

User can define custom position in the style configuration.

screen 2018-01-14 12 20 34 1708x714

If it interesting to someone, then we can improve this draft, otherwise just close it 😉

### CODE REVIEW

File: templates/cassiopeia/index.php [hunk 1]
```diff
@@ -145,6 +149,26 @@
 	</div>
 	<?php endif; ?>
 
+	<?php
+	foreach($params->get('positionsTop', array()) as $position) {
```
Comment (Quy) [inline templates/cassiopeia/index.php:153]: Add a space after `foreach`.

**Comment by brianteeman** [comment]
Nice idea. Right now though it is tied to a specific template. Not sure but if it was a plugin perhaps then it could be used with any

File: templates/cassiopeia/index.php [hunk 2]
```diff
@@ -156,6 +180,25 @@
 		<div class="container-component">
 			<jdoc:include type="modules" name="main-top" style="cardGrey" />
 			<jdoc:include type="message" />
+			<?php
+			foreach($params->get('positionsMiddle', array()) as $position) {
```
Comment (Quy) [inline templates/cassiopeia/index.php:184]: Add space after `foreach` and move `{` to its own line.

File: templates/cassiopeia/index.php [hunk 3]
```diff
@@ -169,6 +212,26 @@
 
 	</div>
 
+	<?php
+	foreach($params->get('positionsBottom', array()) as $position) {
```
Comment (Quy) [inline templates/cassiopeia/index.php:216]: Same here.

**Comment by Fedik** [comment]
> Nice idea. Right now though it is tied to a specific template. Not sure but if it was a plugin perhaps then it could be used with any

yeap, it only to the template which has this options

**Comment by ciar4n** [comment]
Personally I'm still in favor of having a solution in core rather than in the template. By having the solution in core we can set a standard regardless of the template and allow better integration within the Joomla workflow.

Having said that if we can not accommodate for a solution in core then this is the next best thing.

**Comment by brianteeman** [comment]
Thats why I was suggesting it should be a plugin (if possible) so it could be "in core"

**Comment by Fedik** [comment]
> ... we can set a standard regardless of the template ...

That what I actually tried. Make the standard more "clear".
Tried to say the template developer should care about the positions.

> By having the solution in core we can set a standard regardless of the template and allow better integration within the Joomla workflow.

how module helps here? it just add more confusion. because you still need to have "that position" in template to display at least "module block" 😉 

> Thats why I was suggesting it should be a plugin (if possible) so it could be "in core"

It would be better of course, but for now I have no idea how to do it. We can extend the form of the template params with plugin, but rendering of a positions is up to a template developer.
This need to think.

For now I tried to keep it simple, just to show an example for other how to allow to make custom positions in their template. 
In which direction to go 😄

**Comment by Chacapamac** [comment]
Just sending this for Ideas. Don't know if relevant?
I use **Gantry** for a long time, their template system is, for me, extremely flexible and easy to understand.
Even compare to the new kids on the block (I try), this is a powerull and super easy way to build web site.
 
gantry

**Comment by brianteeman** [comment]
related https://github.com/joomla/joomla-cms/pull/19062 ??

**Pull request closed without merging**